### ISSUE TITLE
core dump with automatic async dist connection

### ISSUE THREAD

**Issue by max-au** [erlang/otp#4964]
VM dumps the core (or aborts in a debug build) when dist_auto_connect it enabled (which is the default).

This appears to be a race condition, when erl_node_tables collects garbage (every 120 sec by default, but can be configured down to 1 second), and `{auto_connect, Node, DHandle}` messages are still in the net_kernel message queue, pointing to DistEntry that has been already freed.

**To Reproduce**

Use emulator with option "+zdntgc 1" to simplify the repro

```
-module(dist_crash).
-export([main/1]).

-mode(compile).

main([]) ->
    inet_db:set_lookup([file]),
    Nodes = [list_to_atom("none@host" ++ integer_to_list(Seq)) || Seq <- lists:seq(1, 200)],
    Self = self(),
    [spawn(fun () -> sender(Self, Nodes) end) || _ <- lists:seq(1, 200)],
    count(0).

-define(PRINT_EVERY, 100).

count(Total) ->
    receive
        count when Total rem ?PRINT_EVERY =:= 0 ->
            io:format("Total attempts: ~bk~n", [Total div ?PRINT_EVERY]),
            count(Total + 1);
        count ->
            count(Total + 1)
    end.

sender(Main, Nodes) ->
    [{none, Node} ! msg || Node <- Nodes],
    Main ! count,
    sender(Main, Nodes).
```

**Comment by sverker** [comment]
@max-au

- What OTP version did you reproduce with above test code?
- What OS and hardware?
- How long did you run for it to crash?

**Comment by sverker** [comment]
Just got a crash on OTP-24.0.2, Ubuntu 18.04.5 on Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz with 8 cpu cores:
```
> cerl -debug +zdntgc 1 -sname kalle
1> dist_crash:main([]).
...
Total attempts: 279k
Total attempts: 280k
erts_refc_read(): Bad refc found (refc=0 < 1)!
Aborted (core dumped)
```
Does that look familiar?

**Comment by max-au** [comment]
Yes. It also reproduces on Mac OS X (Intel CPU), but takes considerably less time (circa 50-60k attempts) if I run it with `+S 4:4` to limit normal schedulers.
Assertion is only printed in debug build, non-debug dumps core in allocation functions stack frames.

**Comment by sverker** [comment]
Finally I think I found the race. Proposed fix in #5015.

**Comment by sverker** [comment]
Fix #5015 merged to maint for OTP-24.1, also planned for next maint-23 and maint-22 patch releases.

Thanks @max-au for the reproducing test code. With it, finding the race was practically possible.
